File: typescript/src/Auth/AuthManager.ts
/**
 * Zirve Auth Manager — Keycloak Integration.
 *
 * Provides token introspection, user info, and role verification using the Keycloak OIDC endpoint.
 */
export class AuthManager {
  private config: Record<string, any>;
  private baseUrl: string;
  private realm: string;

  constructor(config: Record<string, any>) {
    this.config = config;
    this.baseUrl = (config.url || 'http://keycloak.zirve-infra.svc.cluster.local').replace(/\/$/, '');
    this.realm = config.realm || 'zirve';
  }

  /**
   * Introspect an OAuth2 / OIDC access token.
   * Checks if the token is active and returns its claims.
   */
  public async verifyToken(token: string): Promise<Record<string, any>> {
    const clientId = this.config.client_id || 'zirve-backend';
    const clientSecret = this.config.client_secret || '';

    const params = new URLSearchParams();
    params.append('token', token);
    params.append('client_id', clientId);
    if (clientSecret) {
      params.append('client_secret', clientSecret);
    }

    const response = await fetch(`${this.baseUrl}/realms/${this.realm}/protocol/openid-connect/token/introspect`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: params.toString(),
    });

    if (!response.ok) {
      throw new Error(`Keycloak introspection failed with status: ${response.status}`);
    }

    const data = await response.json();
    return data;
  }

  /**
   * Check if a token has a specific realm role.
   */
  public async hasRole(token: string, roleName: string): Promise<boolean> {
    const claims = await this.verifyToken(token);
    if (!claims.active) {
      return false;
    }

    const roles = claims.realm_access?.roles || [];
    return roles.includes(roleName);
  }

  /**
   * Fetch the user info from Keycloak based on the access token.
   */
  public async getUser(token: string): Promise<Record<string, any>> {
    const response = await fetch(`${this.baseUrl}/realms/${this.realm}/protocol/openid-connect/userinfo`, {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });

    if (!response.ok) {
      throw new Error(`Keycloak userinfo failed with status: ${response.status}`);
    }

    return await response.json();
  }

  /**
   * Basic health check.
   */
  public async health(): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/realms/${this.realm}/.well-known/openid-configuration`, {
        method: 'GET',
      });
      return response.ok;
    } catch {
      return false;
    }
  }
}

File: dotnet/Zirve.Sdk/Auth/AuthManager.cs
using System;
using System.Net.Http;
using System.Net.Http.Json;
using System.Text.Json;
using System.Threading.Tasks;
using System.Collections.Generic;
using Zirve.Sdk.Config;

namespace Zirve.Sdk.Auth;

/// <summary>
/// Zirve Auth Manager — Keycloak Integration.
/// Validates tokens, checks roles, and retrieves user info via OIDC endpoints.
/// </summary>
public class AuthManager
{
    private readonly HttpClient _httpClient;
    private readonly ConfigManager _config;
    private readonly string _baseUrl;
    private readonly string _realm;

    public AuthManager(HttpClient httpClient, ConfigManager configManager)
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        _config = configManager ?? throw new ArgumentNullException(nameof(configManager));
        
        var cfg = _config.Module("auth");
        _baseUrl = (cfg.GetValueOrDefault("url", "http://keycloak.zirve-infra.svc.cluster.local") ?? "").TrimEnd('/');
        _realm = cfg.GetValueOrDefault("realm", "zirve") ?? "zirve";
    }

    /// <summary>
    /// Introspect an OAuth2 / OIDC access token.
    /// </summary>
    public async Task<JsonElement> VerifyTokenAsync(string token)
    {
        var cfg = _config.Module("auth");
        var clientId = cfg.GetValueOrDefault("client_id", "zirve-backend");
        var clientSecret = cfg.GetValueOrDefault("client_secret", "");

        var content = new FormUrlEncodedContent(new[]
        {
            new KeyValuePair<string, string>("token", token),
            new KeyValuePair<string, string>("client_id", clientId!),
            new KeyValuePair<string, string>("client_secret", clientSecret!)
        });

        var response = await _httpClient.PostAsync($"{_baseUrl}/realms/{_realm}/protocol/openid-connect/token/introspect", content);
        response.EnsureSuccessStatusCode();

        var json = await response.Content.ReadAsStringAsync();
        return JsonSerializer.Deserialize<JsonElement>(json);
    }

    /// <summary>
    /// Check if the introspected token is active and contains the requested realm role.
    /// </summary>
    public async Task<bool> HasRoleAsync(string token, string roleName)
    {
        var claims = await VerifyTokenAsync(token);
        
        if (!claims.TryGetProperty("active", out var active) || !active.GetBoolean())
        {
            return false;
        }

        if (claims.TryGetProperty("realm_access", out var realmAccess) && 
            realmAccess.TryGetProperty("roles", out var roles) &&
            roles.ValueKind == JsonValueKind.Array)
        {
            foreach (var role in roles.EnumerateArray())
            {
                if (role.GetString() == roleName) return true;
            }
        }

        return false;
    }

    /// <summary>
    /// Get OIDC UserInfo.
    /// </summary>
    public async Task<JsonElement> GetUserAsync(string token)
    {
        var request = new HttpRequestMessage(HttpMethod.Get, $"{_baseUrl}/realms/{_realm}/protocol/openid-connect/userinfo");
        request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", token);

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        var json = await response.Content.ReadAsStringAsync();
        return JsonSerializer.Deserialize<JsonElement>(json);
    }

    public async Task<bool> HealthAsync()
    {
        try
        {
            var response = await _httpClient.GetAsync($"{_baseUrl}/realms/{_realm}/.well-known/openid-configuration");
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }
}
